Add blosc2

FrancescAlted opened the pull request:
### Description

This adds the python-blosc2 package (https://github.com/Blosc/python-blosc2).

FrancescAlted commented:
Pinging @agriyakhetarpal

agriyakhetarpal reviewed (commented):
Thanks, @FrancescAlted! This looks like a good start. Is there a compile-time setting that we might need to patch in the Blosc2 sources? I ask because I notice that almost all of the functionality will need to set ``nthreads=1`` everywhere (see https://pyodide.org/en/stable/usage/wasm-constraints.html#included-but-not-working-modules).

FrancescAlted commented on packages/blosc2/meta.yaml:
```diff
@@ -0,0 +1,33 @@
+  license: BSD-3-Clause
+extra:
+  recipe-maintainers:
+    - FrancescAlted
```
Done

FrancescAlted commented on packages/blosc2/meta.yaml:
```diff
@@ -0,0 +1,33 @@
+    - numpy
+    #- ndindex  # TODO: this is not present
+    - msgpack
+    #- platformsdir  # TODO: this is not present
```
Yes, it is `platformdirs` indeed.  Although this is a pure python package, so probably we don't need to package it in pyodide, right? I did some testing with pyodide console, and this seems to be imported flawlessly.

FrancescAlted commented on packages/blosc2/meta.yaml:
```diff
@@ -0,0 +1,33 @@
+  run:
+    # Dependencies that are needed to run the package
+    - numpy
+    #- ndindex  # TODO: this is not present
```
Same here: it seems to me that it is not necessary to add a recipe for it (I might be wrong though).

FrancescAlted commented on packages/blosc2/meta.yaml:
```diff
@@ -0,0 +1,33 @@
+    - requests
+build:
+  script: |
+    export WITH_OPTIM=OFF   # TODO: not sure how to pass -DWITH_OPTIM to cmake (used indirectly by scikit-build-core)
```
Great, that works flawlessly.

FrancescAlted commented:
For me, this is working already; there is the open question on whether should we add recipes for `ndindex` and `platformdirs` though.

Regarding the WASM build and test for CI, it is already working here: https://github.com/Blosc/python-blosc2/actions/workflows/wasm.yml. Is that ok?

FrancescAlted commented:
Also, there is the question of the package name. In github, the name of the project is `python-blosc2`, but in PyPI is `blosc2`. I am using the PyPI convention here. Would that be ok?

FrancescAlted commented:
> Thanks, @FrancescAlted! This looks like a good start. Is there a compile-time setting that we might need to patch in the Blosc2 sources? I ask because I notice that almost all of the functionality will need to set `nthreads=1` everywhere (see https://pyodide.org/en/stable/usage/wasm-constraints.html#included-but-not-working-modules).

Yeah, python-blosc2 already detects when it is in a `wasm32` platform, and never raises `nthreads` to more than 1.

FrancescAlted commented:
I see.  Ok, I'll try to send PRs for `ndindex` and `platformdirs` then.  I'll keep you informed.

agriyakhetarpal commented:
Please feel free to include them within the same PR if you like. Our CI takes a long time to finish (>=2 hours), and I guess one PR would speed up the process for you, rather than having to deal with three PRs.

ryanking13 commented on pyodide-build:
Could you please revert pyodide-build submodule update? I don't think it is necessary in this PR.

FrancescAlted replied:
Ups. Reverted updates.

FrancescAlted commented:
> Please feel free to include them within the same PR if you like. Our CI takes a long time to finish (>=2 hours), and I guess one PR would speed up the process for you, rather than having to deal with three PRs.

Ups, I already went ahead and sent different PRs for that:

* `platformdirs` https://github.com/pyodide/pyodide/pull/5485
* `ndindex` https://github.com/pyodide/pyodide/pull/5487

Should I close them?

agriyakhetarpal commented:
> > Please feel free to include them within the same PR if you like. Our CI takes a long time to finish (>=2 hours), and I guess one PR would speed up the process for you, rather than having to deal with three PRs.
> 
> Ups, I already went ahead and sent different PRs for that:
> 
> * `platformdirs` [Packaging platformdirs #5485](https://github.com/pyodide/pyodide/pull/5485)
> * `ndindex` [Ndindex #5487](https://github.com/pyodide/pyodide/pull/5487)
> 
> Should I close them?

Yes, please do, if you feel like it! If this PR is merged with all three recipes, it would be easier to know why the other two pure Python packages were added.

agriyakhetarpal commented:
The Pillow build looks flaky; I've re-triggered it. In the meantime, I'll try to build it locally.

Could you please add a test (or a couple of them) for each package? `platformdirs` and `ndindex` should be easy candidates, and you may use your judgement for `blosc2` as the core developer as to what subset of the upstream test suite can be included. I think we can merge this PR after that.

FrancescAlted commented:
> The Pillow build looks flaky; I've re-triggered it. In the meantime, I'll try to build it locally.

Pillow?  I don't think we need pillow in blosc2.

> Could you please add a test (or a couple of them) for each package? `platformdirs` and `ndindex` should be easy candidates, and you may use your judgement for `blosc2` as the core developer as to what subset of the upstream test suite can be included. I think we can merge this PR after that.

I have added some tests for blosc2; although not that much, they exercise a good part of the blosc2 machinery.  Do you think the format is ok?  I don't have much time right now, but tomorrow I'll try to come back to this; at any rate, if you want to add tests for patformdirs and ndindex, feel free (I am not familiar with their test suites).

agriyakhetarpal commented:
BTW, it's quite surprising – it is only now that I got to know that `ndindex` isn't a pure Python package!

FrancescAlted commented:
> We don't – it's just that Pillow arrives in our build graph before `blosc2` does, and we build every package on every commit (yes, it's messy). Sometimes, unrelated failures show up, like this one.

I see.

> Yes, the format is looking great for a start, but you'll also need the `@run_in_pyodide` decorator here from the `pytest-pyodide` plugin, and every test function needs `selenium` as its first argument. This is purely due to the way we've set up our test suite. I was able to build `blosc2` locally just fine, so it's just a matter of ensuring the right format. I think I can add a test each for `platformdirs` and `ndindex`, though I'll wait for you to push your changes before I push mine so that I don't cause trouble for your local branch. :)

I have pushed a good set of tests for blosc2 that apparently run fine in CI.  Thanks for all the help.

agriyakhetarpal commented:
Thank you! We also need `import blosc2` and `import numpy as np` statements inside all test functions, as they all run in an isolated manner. We also need `"numpy"` in the `@run_in_pyodide` decorator where applicable so that it knows to fetch the package. I fixed these for you, and I also added tests for `platformdirs` and `ndindex`.

There are two failing tests in `blosc2`. Could you please take a look at them? I tried to compare the expected values in `assert dest2 == bytes_obj` by hand, but they are too big to do so. The rest of the cases for this test are passing.

```python
FAILED blosc2/test_blosc2.py::compress2[node-641091-cparams1-dparams1-True] - AssertionError
FAILED blosc2/test_blosc2.py::compress2[node-641091-cparams1-dparams1-False] - AssertionError
```

FrancescAlted commented:
> Thank you! We also need `import blosc2` and `import numpy as np` statements inside all test functions, as they all run in an isolated manner. We also need `"numpy"` in the `@run_in_pyodide` decorator where applicable so that it knows to fetch the package. I fixed these for you, and I also added tests for `platformdirs` and `ndindex`.
> 
> There are two failing tests in `blosc2`. Could you please take a look at them? I tried to compare the expected values in `assert dest2 == bytes_obj` by hand, but they are too big to do so. The rest of the cases for this test are passing.
> 
> ```python
> FAILED blosc2/test_blosc2.py::compress2[node-641091-cparams1-dparams1-True] - AssertionError
> FAILED blosc2/test_blosc2.py::compress2[node-641091-cparams1-dparams1-False] - AssertionError
> ```

Cool! that was an oversight on my side. This test was using 4 threads, but this is not possible in pyodide.  This should have been fixed now.

agriyakhetarpal commented on pyodide-build:
Could you please revert this change? I think it might have come up in a commit somewhere.

FrancescAlted replied:
Should be fixed.  Sorry about that.

agriyakhetarpal reviewed (approved):
Thanks a lot for collaborating with me on this, @FrancescAlted! Just one change and I think we should be good to go now.

hoodmane commented:
Thanks @FrancescAlted!

agriyakhetarpal commented:
The 0.28 release needs the unvendoring of recipes to happen and Python 3.13. I think it can come up in April. #5495 lists a few PRs that need reviews before that happens. You may use [a custom Pyodide distribution](https://jupyterlite.readthedocs.io/en/stable/howto/pyodide/pyodide.html) or [ship a wheel built out-of-tree](https://jupyterlite.readthedocs.io/en/stable/howto/pyodide/wheels.html) if you intend to use JupyterLite, or try it out on[ the "latest" Pyodide console](https://pyodide.org/en/latest/console.html) from the main branch (which isn't intended for production).
